### DISCUSSION TITLE
RFC-011 Supports `data:text/javascript` protocol

### DISCUSSION THREAD

**Discussion by hyf0** [web-infra-dev/rspack#457] [RFC]
- Feature Name: Supports `data:text/javascript` protocol
- Start Date: 2022-07-19
- RFC PR: WIP
- Issue: #456 

# Summary
[summary]: #summary

Supports bundling code like'
```ts
import x from "data:text/javascript,export default 42"
console.log('x:',x);
```

# Motivation
[motivation]: #motivation

It seems in the spec, but I forgot which one.
__It's one way to deal with virtual files__.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Let's say you have an idea that you want to inject some runtime code in bundling.

The idea requests you to deal with virtual modules.

virtual modules are javascript files that do not exist on the hard disks.

See [the way](https://vitejs.dev/guide/api-plugin.html#virtual-modules-convention) how vite does it.
```ts
export default function myPlugin() {
  const virtualModuleId = 'virtual:my-module'
  const resolvedVirtualModuleId = '\0' + virtualModuleId

  return {
    name: 'my-plugin', // required, will show up in warnings and errors
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return `export const msg = "from virtual module"`
      }
    }
  }
}

// foo.js

```ts
import { msg } from 'virtual:my-module'
```

You could see that vite intercepts the procedure of reading files from disk.

And that's the way using `data:text/javascript` protocol.

```ts
import { msg } from 'data:text/javascript,export const msg = "from virtual module"'
```

RSpack is ok to use Vite's way, but it has some problems. Such as the `virtualModuleId` will be exposed to other plugins. Since Vite's way is only a conversion not in the spec, people don't have a consistent solution, which could lead to some problems.

# Reference-level explanation

- The implementation will be looks like Webpack, so you could see the part of the prior art first.

## How to handle importing modules by relative paths in the `data:` protocol.
  
For example, attempting to load `./foo` from `data:text/javascript,import "./foo"`;
  
I tend to raise an error just like node.js did.

# Drawbacks
[drawbacks]: #drawbacks

It's in the spec, no significant reasons we shouldn't do it.
# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives
- Why is this design the best in the space of possible designs?
  
  It's in the spec.
- What other designs have been considered and what is the rationale for not choosing them?
  
  See the [Motivation](#Motivation) part.

# Prior art
[prior-art]: #prior-art
## Webpack

- Webpack supports sheme and has a special hook `resolveForScheme` [when resolving a module in `NormalModuleFactory`](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModuleFactory.js#L685-L700)
- `import { msg } from 'data:text/javascript,export const msg = "from virtual module"'` will be parsed to two `Dependency`s.
	- `HarmonyImportSideEffectDependency`
	- `HarmonyImportSpecifierDependency`
![image](https://user-images.githubusercontent.com/49502170/182142290-1b120d38-9601-43f7-82bd-d23f38f27f0c.png)

![image](https://user-images.githubusercontent.com/49502170/182142309-7236316b-7047-48b0-b9c4-abce312643eb.png)

- Webpack will try resolved to a `NormalModule` using the two `Dpendency`s(as a whole) in [`NormalModuleFactory#create`](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModuleFactory.js#L740)
- Webpack will [read the request](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModuleFactory.js#L359) [property in HarmonyImportSideEffectDependency](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModuleFactory.js#L745)
- [If a `ModuleRequest` has scheme](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModuleFactory.js#L359), Webpack will call [`resolveForScheme`](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModuleFactory.js#L684-L700) hook and the [DataURI plugin](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/schemes/DataUriPlugin.js) will use the hook to return the information of `data:...` to Webpack.
- In the end, we got a `NormalModule`.
- while `NormalModule` rendering,  `NormalModule` will [call `readResource`](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModule.js#L824-L829)  hook to get original source code.
- The DataURI plugin will use [`readResourceForScheme`](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/schemes/DataUriPlugin.js#L48-L50) hook to provide source code.
	- [`readResourceForScheme` hook is wrapper of `readResource` hook](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/NormalModule.js#L239-L240), which will be deprecated in Webpack 6.
- The rest of the process is just like normal imported module.

### Related

- [DataUriPlugin](https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/schemes/DataUriPlugin.js)

# Unresolved questions

Nope

# Future possibilities

[future-possibilities]: #future-possibilities

We should also supports `data:text,html` and `data:text,css`.

**Comment by hardfist** [comment]
some related discussion in webpack https://github.com/webpack/webpack/issues/6437#issuecomment-920341588, it's the recommend way to deal virtual module in webpack, @sodatea and @antfu may have better ideas

**Reply by hardfist** [reply]
see 
* https://nodejs.org/api/esm.html#data-imports, 
* https://www.rfc-editor.org/rfc/rfc2397#section-2
* https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs

and actually data-url scheme is not only used in javascript but also used in html and css, it's the first class citizen, so we may need to handle it both in javascript、css and html

**Reply by byted-meow** [reply]
[NodeJS's data protocol](https://nodejs.org/api/esm.html#data-imports) doesn't have a resolving context so relative specifiers like `./index.js` will fail to resolve. Do we need to patch it to support relative specifiers?

**Reply by hyf0** [reply]
I tend to the way of nodejs.

**Comment by hyf0** [comment]
The RFC now is blocked by the unimplemented new architecture of RSpack.
